### PR TITLE
Fix crash with invalid proxy configuration

### PR DESCRIPTION
**Pull request by ewertons**
This replaces the original PR https://github.com/Azure/azure-c-shared-utility/pull/564
Unit tests were added to this new PR.

### CODE REVIEW

File: tests/http_proxy_io_ut/http_proxy_io_ut.c [hunk 1]
```diff
@@ -1575,6 +1575,70 @@ TEST_FUNCTION(http_proxy_io_retrieve_options_calls_the_underlying_retrieve_optio
     http_proxy_io_get_interface_description()->concrete_io_destroy(http_io);
 }
 
+#define OPTION_UNDERLYING_IO_OPTIONS "underlying_io_options"
+
+/* GitHub PR 564 */
```
Comment (danewalton) [inline tests/http_proxy_io_ut/http_proxy_io_ut.c:1580]: these labels necessary?

File: build/.vsts-ci.yml [hunk 2]
```diff
@@ -5,7 +5,7 @@ resources:
 jobs:
 - job: checksubmodule
   pool:
-    name: Hosted Ubuntu 1604
+    vmImage: 'ubuntu-20.04'
```
Comment (danewalton) [inline build/.vsts-ci.yml:8]: this is moving on from self hosted right?

**Review by danewalton — APPROVED** [review]

File: tests/http_proxy_io_ut/http_proxy_io_ut.c (hunk 1 above)
Comment (ewertons) [reply tests/http_proxy_io_ut/http_proxy_io_ut.c:1580]: Oh, the label yes, but the define should be moved to the top. Thanks.

File: build/.vsts-ci.yml (hunk 2 above)
Comment (ewertons) [reply build/.vsts-ci.yml:8]: Yes, correct, we are using Azure DevOps agents for that.

**Pull request merged**